linux/mcompile/keymap/u16.cpp

```diff
@@ -0,0 +1,329 @@
+			s += (48 + temp);
+		else
+			s += (87 + temp);
+		num = num / 16;
+	}
+	reverse(s.begin(), s.end());
+	return s;
+}*/
+
+const KMX_WCHAR *  u16ncat(KMX_WCHAR *dst, const KMX_WCHAR *src, size_t max) {
+  KMX_WCHAR* o = dst;
+  dst = (KMX_WCHAR*) u16chr(dst, 0);
+	//max -= (dst-o);
+  while (*src && max > 0) {
+    *dst++ = *src++;
+    max--;
+  }
+	if(max > 0)
+  	*dst = 0;
+  return o;
+}
+
+const KMX_WCHAR* u16rchr_slash(KMX_WCHAR const* Name)
+{
+  const KMX_WCHAR* cp = NULL;
+  cp = u16rchr(Name, '\\');
+  if (cp == NULL)
+    cp = u16rchr(Name, '/');
+  return cp;
+}
+/*
+KMX_CHAR* strrchr_slash(KMX_CHAR* Name)
+{
+  KMX_CHAR* cp = NULL;
+  cp = strrchr(Name, '\\');
+  if (cp == NULL)
+    cp = strrchr(Name, '/');
+  return cp;
+}
+*/
+// u16rchr returns last occurence of ch in p; It returns NULL  if ch = '\0' and NULL if ch is not found
+const KMX_WCHAR* u16rchr(const KMX_WCHAR* p, KMX_WCHAR ch) {
+  const KMX_WCHAR* p_end = p + u16len(p) - 1;
+
+	if (ch == '\0')	return p_end + 1;
+	while (p_end >= p) {
+		if (*p_end == ch) return p_end;
+		p_end--;
+	}
+	return NULL;
+}
+
+const KMX_WCHAR *  u16chr(const KMX_WCHAR *p, KMX_WCHAR ch) {
+  while (*p) {
+    if (*p == ch) return p;
+    p++;
+  }
+  return ch == 0 ? p : NULL;
+}
+
+const KMX_WCHAR *  u16cpy(KMX_WCHAR *dst, const KMX_WCHAR *src) {
+  KMX_WCHAR *o = dst;
+  while (*src) {
+    *dst++ = *src++;
+  }
+  *dst = 0;
+  return o;
+}
+
+const KMX_WCHAR *  u16ncpy(KMX_WCHAR *dst, const KMX_WCHAR *src, size_t max) {
+  KMX_WCHAR *o = dst;
+  while (*src && max > 0) {
+    *dst++ = *src++;
+    max--;
+  }
+  while(max > 0) {
+    *dst++ = 0;
+    max--;
+  }
+  return o;
+}
+
+size_t   u16len(const KMX_WCHAR *p) {
+  int i = 0;
+  while (*p) {
+    p++;
+    i++;
+  }
+  return i;
+}
+
+int   u16cmp(const KMX_WCHAR *p, const KMX_WCHAR *q) {
+  while (*p && *q) {
+    if (*p != *q) return *p - *q;
+    p++;
+    q++;
+  }
+  return *p - *q;
+}
+
+int   u16nicmp(const KMX_WCHAR *p, const KMX_WCHAR *q, size_t count) {
+  while (*p && *q && count) {
+    if (toupper(*p) != toupper(*q)) return *p - *q;
+    p++;
+    q++;
+    count--;
+  }
+  if (count)
+    return *p - *q;
+  return 0;
+}
+
+int   u16icmp(const KMX_WCHAR *p, const KMX_WCHAR *q) {
+  while (*p && *q) {
+    if (toupper(*p) != toupper(*q)) return *p - *q;
+    p++;
+    q++;
+  }
+  return *p - *q;
+}
+
+int   u16ncmp(const KMX_WCHAR *p, const KMX_WCHAR *q, size_t count) {
+  while (*p && *q && count) {
+    if (*p != *q) return *p - *q;
+    p++;
+    q++;
+    count--;
+  }
+  if (count)
+    return *p - *q;
+  return 0;
+}
+
+KMX_WCHAR * u16tok(KMX_WCHAR *p,  KMX_WCHAR ch,  KMX_WCHAR **ctx) {
+  if (!p) {
+    p = *ctx;
+    if (!p) return NULL;
+  }
+
+  KMX_WCHAR *q = p;
+  while (*q && *q != ch) {
+    q++;
+  }
+  if (*q) {
+    *q = 0;
+    q++;
+    while (*q == ch) q++;
+    *ctx = q;
+  }
+  else {
+    *ctx = NULL;
+  }
+  return p;
+}
+
+KMX_WCHAR * u16tok(KMX_WCHAR* p,  KMX_WCHAR* delim, KMX_WCHAR** ctx) {
```

Comment:
I'm not sure what this method is supposed to do exactly. If `delim` is supposed to be a string then it doesn't work as expected. If it is supposed to be an array of delimiters then the variable should be named differently, e.g. `delimiters` or `tokens`. Again, unit tests would clarify this.